allennlp/nn/util.py

```diff
@@ -402,9 +403,99 @@ def inner_device_mapping(storage: torch.Storage, location) -> torch.Storage:  #
+    just use something like ``torch.cat([x_tensor, y_tensor, x_tensor * y_tensor])``.  This
+    function adds some complexity that is only necessary if you want the specific combination used
+    to be `configurable`.
+
+    If you want to do any element-wise operations, the tensors involved in each element-wise
+    operation must have the same shape.
+
+    This function also accepts ``x`` and ``y`` in place of ``1`` and ``2`` in the combination
+    string.
+    """
+    combination = combination.replace('x', '1').replace('y', '2')
+    combinations = combination.split(',')
+    combined_tensor = _get_combination(combinations[0], tensors)
+    for piece in combinations[1:]:
+        to_concatenate = _get_combination(piece, tensors)
+        combined_tensor = torch.cat([combined_tensor, to_concatenate], dim=-1)
+    return combined_tensor
+
+def _get_combination(combination: str, tensors: List[torch.Tensor]) -> torch.Tensor:
+    if combination.isdigit():
+        index = int(combination) - 1
+        return tensors[index]
+    else:
+        if len(combination) != 3:
+            raise ConfigurationError("Invalid combination: " + combination)
+        first_tensor = _get_combination(combination[0], tensors)
+        second_tensor = _get_combination(combination[2], tensors)
+        operation = combination[1]
+        if operation == '*':
+            return first_tensor * second_tensor
+        elif operation == '/':
+            return first_tensor / second_tensor
+        elif operation == '+':
+            return first_tensor + second_tensor
+        elif operation == '-':
+            return first_tensor - second_tensor
+        else:
+            raise ConfigurationError("Invalid operation: " + operation)
+
+
+def get_combined_dim(combination: str, tensor_dims: List[int]) -> int:
+    """
+    For use with :func:`combine_tensors`.  This function computes the resultant dimension when
+    calling ``combine_tensors(combination, tensors)``, when the tensor dimension is known.  This is
+    necessary for knowing the sizes of weight matrices when building models that use
+    ``combine_tensors``.
+
+    Parameters
+    ----------
+    combination : ``str``
+        A comma-separated list of combination pieces, like ``"1,2,1*2"``, specified identically to
+        ``combination`` in :func:`combine_tensors`.
+    tensor_dims : ``List[int]``
+        A list of tensor dimensions, where each dimension is from the `last axis` of the tensors
+        that will be input to :func:`combine_tensors`.
+    """
+    combination = combination.replace('x', '1').replace('y', '2')
+    return sum([_get_combination_dim(piece, tensor_dims) for piece in combination.split(',')])
+
+
+def _get_combination_dim(combination: str, tensor_dims: List[int]) -> int:
+    if combination.isdigit():
+        index = int(combination) - 1
+        return tensor_dims[index]
+    else:
+        if len(combination) != 3:
```

Comment:
this seems to assume only single digits, maybe that's a good assumption in practice, but it seems dangerous to rely on it

Reply:
Concatenating ten or more tensors at a time seems pretty implausible.  I'll raise an error if you try to use it with more than is supported - that's easier than making the logic robust to multiple digits, and we'll almost certainly never need more than 9.